### PR TITLE
Add an "--all" option to the got command

### PR DESCRIPTION
**Pull request by gordonmessmer**
## Description

This change adds a "got-all" command which expands on the existing "got" command by providing data about relocations in mapped shared object files in addition to the relocations specific to the main executable.

Particularly for auditing purposes, users may be interested in the state of all relocations, not only those for the primary executable file.

-  [x] My code follows the code style of this project.
-  [x] My change includes a change to the documentation, if required.
-  [x] If my change adds new code, [adequate tests](docs/testing.md) have been added.
-  [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by gordonmessmer** [comment]
This draft is functional, but incomplete.  It does not include documentation or tests.  In its current condition, it exists only to ask whether the maintainers are interested in expanding the "got" command to provide information about relocations in mapped shared object files or not.  And, if so, what the output from that command might look like.  The current implementation merely prints relocations for all files, without indicating which file's relocations are being printed.

I'm also not sure whether such a command needs to accept a filter, allowing the user to print relocations for specific files, other than the primary executable file.

Please let me know what you think, and whether or not I should finish work on this command.

**Comment by Grazfather** [comment]
Can you please show its use? Couldn't you just add this as an argument to the existing got command?

**Comment by gordonmessmer** [comment]
I can show its output, but it tends to be _quite_ long.  I'll try attaching it as a file.

The existing got command will show the state of relocatable symbols present in the primary binary executable.  That's not all of the relocations, though.  For example, the `login` binary is linked to libpam.so.0 on GNU/Linux systems, and libpam uses dlopen() to load libraries from `/usr/lib64/security` (or a similar path).  The existing `got` command will not show any instance of dlopen, because there's a separate offset table for each ELF library in addition to the offset table for the main binary.

Arguments to the current `got` command reduce its output by filtering for matching symbols.  The proposed `got-all` command makes the output much longer, by printing all of the offset tables for each mapped shared object in addition to the binary executable.

**Comment by gordonmessmer** [comment]
Attaching got-all and got output for the login process

[got-all.txt](https://github.com/hugsy/gef/files/15382597/got-all.txt)
[got.txt](https://github.com/hugsy/gef/files/15382598/got.txt)

**Comment by Grazfather** [comment]
Yes I understand, but we could add an argument to have it run got 'deeper'. I see this as a recursive GOT command, but with depth one. We could add a `--all` argument, or even a `--depth` argument that limits how deep it looks into libraries' libraries.

**Comment by gordonmessmer** [comment]
You could do that, but since the existing behavior is for all arguments to act as filters, the change would not be fully backward compatible, which is largely why I didn't pursue that path.

If that's your preference, I can continue developing in that direction.

**Comment by Grazfather** [comment]
I believe it would be, but maybe it would be breaking searching for symbols that contain `-all`. @hugsy care to weigh in?

**Comment by hugsy** [comment]
>  I see this as a recursive GOT command, but with depth one. We could add a `--all` argument, or even a `--depth` argument that limits how deep it looks into libraries' libraries.

I'm with @Grazfather on this, I don't feel like this doesn't deserve its own standalone command, but if you wish to make it so, then I'd suggest moving it to `gef-extras` instead. 
In any case, yes tests and docs are lacking, to explain more usage to users.

**Comment by gordonmessmer** [comment]
I've modified the implementation to use an optional flag to the got command.

The current implementation prints the GOT for each shared object without labeling them, which isn't very user-friendly.  What would be idiomatic for GEF?

File: gef.py
Comment (Grazfather) [inline gef.py:9285]: Let's name this something better. It might not even need to be a separate function, but you might want to still break some of the functionality into helper functions.

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:41]: What can we test that will show up in `got --all` but won't show up without it?

File: gef.py
Comment (gordonmessmer) [reply gef.py:9285]: Do you have any suggestions?

**Comment by Grazfather** [comment]
you could do something like
```python
title = Color.colorify(lib_name, "yellow bold") # idc about the colour much, but should be bold
```
or maybe
```python
title = titlify(lib_name)
```
though this is used more for drawing the context window.

**Comment by gordonmessmer** [comment]
Using titlify:

```
gef➤  got --all malloc
──────────────────────────────────────────── /usr/bin/sleep ────────────────────────────────────────────

GOT protection: Full RelRO | GOT functions: 45
 
[0x55555555bf00] malloc@GLIBC_2.2.5  →  0x7ffff7e64670
───────────────────────────────────────── /usr/lib64/libc.so.6 ─────────────────────────────────────────

GOT protection: Full RelRO | GOT functions: 18
 
[0x7ffff7fa0d68] malloc@@GLIBC_2.2.5  →  0x7ffff7e64670
─────────────────────────────────── /usr/lib64/ld-linux-x86-64.so.2 ───────────────────────────────────

GOT protection: Full RelRO | GOT functions: 0

```

File: tests/commands/got.py
Comment (Grazfather) [reply tests/commands/got.py:43]: These checks should make sure that they show up UNDER the libc part.

You could get the res, assert `/libc` is in there, split on it, and then make sure `printf` and `strcpy` are on the second part.

**Comment by Grazfather** [comment]
That looks pretty good to me.

File: tests/commands/got.py
Comment (gordonmessmer) [reply tests/commands/got.py:43]: Split doesn't really work, because it would return two sections: everything before libc and everything after.  AFAIK, the order isn't guaranteed, so the program binary could be in either one.

(Also, printf and strcpy are expected in the program's section, but not in libc's, because those aren't relocatable symbols in libc.)

If you'd like, I can rewrite this to parse the sections...  I'll upload a commit that does that.  If you don't like it, I'll drop that commit and roll back to the current state.

**Comment by gordonmessmer** [comment]
I'm happier with this test setup.

There's still the matter of the do_invoke_for function name, if that's important to you.  Otherwise, if you are happy with this, I can squash it before merging...

File: docs/commands/got.md
Comment (Grazfather) [inline docs/commands/got.md:9]:
```suggestion
gef➤ got [--all] [filters]
```
or similar

File: docs/commands/got.md
Comment (Grazfather) [inline docs/commands/got.md:12]:
```suggestion
`--all` Print the GOT for all shared objects in addition to the executable file
```

File: docs/commands/got.md
Comment (Grazfather) [inline docs/commands/got.md:16]: please wrap like the rest

File: gef.py
Comment (Grazfather) [reply gef.py:9285]: `print_got_for`?

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:47]: just return this

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:52]: just return this

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:61]:
```suggestion
                checked_sections.append(checksyms(lines))
```

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:64]:
```suggestion
        checked_sections.append(checksyms(lines))
```

File: tests/commands/got.py
Comment (Grazfather) [inline tests/commands/got.py:59]: I don't really understand what this is doing, can you please add a comment explaining it?

File: docs/commands/got.md
Comment (gordonmessmer) [reply docs/commands/got.md:16]: Can you clarify what "like the rest" means?  markdownlint does have some findings for pre-existing content in this file, but not for the lines that I've added -- that appears to be shorter than the maximum line length.

Comment (Grazfather) [reply docs/commands/got.md:16]: Right, I'm saying you should wrap your lines to match, rather than doing 80 when I think we allow 100 or 120.

**Comment by Grazfather** [comment]
Other than line length nitpicking looks good to me.

File: docs/commands/got.md
Comment (gordonmessmer) [reply docs/commands/got.md:16]: Oh, the github review makes it look like you were referring to line 12, but I think you're actually referring to line 38.  I've un-wrapped that line.

**Comment by gordonmessmer** [comment]
Thank you for your review.  It sounds like this is just about ready, so I've squashed the commits.

File: docs/commands/got.md
Comment (hugsy) [inline docs/commands/got.md:9]: What is `--filters`? I don't see it in the code, care to elaborate?
If it's supposed to be in the code, also document it

File: gef.py
Comment (hugsy) [inline gef.py:9288]: Those are already `str`, no need for calling `str()`

File: gef.py
Comment (hugsy) [inline gef.py:9281]: I dont think this will work in this way when remoting, because `readelf` is checked locally whereas the mapped files will not point to local files. Either exclude the command from running when not local, or make sure the remote files are correctly tested (against the ones in the temporary namespace).

File: gef.py
Comment (hugsy) [inline gef.py:9314]: You can merge those 2 lines in one

**Review by hugsy — CHANGES_REQUESTED** [review]

File: docs/commands/got.md
Comment (gordonmessmer) [reply docs/commands/got.md]: ...that's embarrassing.  Fingers were moving, brain was not.  That was intended to be only 'filters'.

File: gef.py
Comment (gordonmessmer) [reply gef.py:9281]: I'll take another look at this one, tomorrow.

Comment (hugsy) [reply gef.py:9281]: > I'll take another look at this one, tomorrow.

No problem, thanks!
Also quick note, I've flagged your PR to be part of the next release.

File: docs/commands/got.md
Comment (Grazfather) [reply docs/commands/got.md]: I missed this too!

File: gef.py
Comment (Grazfather) [reply gef.py:9281]: for remote does this normally work by downloading the binary offer, e.g. the file object points to the local copy?

Comment (gordonmessmer) [reply gef.py:9281]: So... the list comprehension is looping over Section objects, and Section objects have a `.realpath` property that looks like it's intended to handle paths in the context of a remote session, so using that value seems like maybe it should be sufficient for this context.

But: the `got` command doesn't seem to work with remote sessions in the main branch.  I've checked out 29fb74efa29ec54296016d963724315e743689fa and run a `gef-remote` session to examine a process on a remote host.  It looks like all of the files have been copied to the local host, in `/tmp/tmp8alv77xz/`, but the `got` command fails:

```
(remote) gef➤  got
[!] Command 'got' failed to execute properly, reason: '/usr/libexec/postfix/master' not found/readable, most gef features will not work
```

I'm new to python extensions for gdb, so I'm actually not sure how to print the full exception/backtrace when something in gef fails, which makes it difficult for me to diagnose this further.

If everything else looks good, maybe this would be a good point to merge this PR, and then start a new PR to work on fixing remote use of the `got` command.  (Unless the failure of the pre-existing got command is the result of something I'm doing wrong.)

**Comment by gordonmessmer** [comment]
(Squashed again.)

**Comment by gordonmessmer** [comment]
I've attached a commit that partially fixes the use of the got command with remote debugging.  However, the realpath() property still doesn't work some of the time, because `self.path` reflects the path the appears in /proc/<pid>/maps, but the file is being downloaded to what looks like the path that appears in the output of `ldd` (and that, I would argue, is a bug).  The path that appears in maps is the real path; the path that appears in the output of `ldd` might include a symlink, and isn't canonicalized with `realpath()`

Personally, I think this is a problem that should be solved in a separate PR.  The patch is a little messy, and fixing the got command when remote debugging isn't strictly related to the --all option.  If you agree, I'm happy to back that commit out and open a new PR.

**Comment by gordonmessmer** [comment]
I think that resolving remote use of got will require fixing bug https://sourceware.org/bugzilla/show_bug.cgi?id=23764

The current suggestion is that python code should use realpath instead using `.filename`, but that isn't possible for remote debugging.

I've requested an account for sourceware's bug system, but those require human approval, so it'll take a while to set that up.  And since this looks like a gdb bug, I don't think it's going to be possible to fix remote use of `got --all` in the immediate future.

The patch that I've provided for remote debugging is probably still good, and appropriate, but as I said earlier: messy.

**Comment by Grazfather** [comment]
GDB bugs take years to fix. Does your last path work decently well? We could just add a note to the docs that it doesn't work well with remote sessions and make a note to fix it later.

**Comment by gordonmessmer** [comment]
> GDB bugs take years to fix. 

Yes, that's what I meant by "not in the immediate future."  :)

> Does your last path work decently well? 

It allows the `got` command to work without the `--all` option, and with the `--all` option, libraries which are found in solib search paths that don't contain any symlinks will work.

But on RHEL systems, /lib64 is a symlink to /usr/lib64, so most solibs won't work if the remote host is RHEL or a related system.

I'll move it to a separate PR, though, because I think it bears some discussion about style and consistency.

**Comment by Grazfather** [comment]
Yeah, let's just add a note to the docs about its current state for this PR.

**Review by Grazfather — APPROVED** [review]

**Review by hugsy — APPROVED** [review]
LGTM, good stuff

**Pull request merged**
